pgjdbc/src/main/java/org/postgresql/jdbc/PgStruct.java

```diff
@@ -0,0 +1,190 @@
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import org.postgresql.core.BaseConnection;
+import org.postgresql.core.Oid;
+import org.postgresql.util.GT;
+import org.postgresql.util.PSQLException;
+
+import java.nio.charset.Charset;
+import java.sql.SQLException;
+import java.sql.Struct;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+public class PgStruct implements Struct {
+  // if at least one of \ " ( ) (white space) , is present in the string, we need to quote the attribute value
+  private static final Pattern NEEDS_ESCAPE_PATTERN = Pattern.compile("[\\\"()\\s,]");
+
+  private final String sqlTypeName;
+  private final PgStructDescriptor descriptor;
+  private final Object[] attributes;
+
+  private final TimestampUtils timestampUtils;
+  private final Charset charset;
+
+  public PgStruct(PgStructDescriptor descriptor, Object[] attributes, BaseConnection connection) {
+    this.sqlTypeName = descriptor.sqlTypeName();
+    this.descriptor = descriptor;
+    this.attributes = attributes;
+    this.timestampUtils = connection.getTimestampUtils();
+    this.charset = Charset.forName(connection.getQueryExecutor().getEncoding().name());
+    resolveAttributes();
+  }
+
+  private void resolveAttributes() {
+    for (int i = 0; i < attributes.length; i++) {
+      if (attributes[i] == null) {
+        continue;
+      }
+      try {
+        int oid = descriptor.pgAttributes()[i].oid();
+        if (oid == Oid.DATE) {
+          attributes[i] = timestampUtils.toDate(timestampUtils.getSharedCalendar(null), attributes[i].toString());
+        } else if (oid == Oid.TIME) {
+          attributes[i] = timestampUtils.toTime(timestampUtils.getSharedCalendar(null), attributes[i].toString());
+        } else if (oid == Oid.TIMESTAMP || oid == Oid.TIMESTAMPTZ) {
+          attributes[i] = timestampUtils.toTimestamp(timestampUtils.getSharedCalendar(null), attributes[i].toString());
+        } else {
+          attributes[i] = JavaObjectResolver.tryResolveObject(attributes[i], oid);
+        }
+      } catch (SQLException ignored) {
+        // skip date time conversions that fail
+      }
+    }
+  }
+
+  @Override
+  public String getSQLTypeName() {
+    return sqlTypeName;
+  }
+
+  @Override
+  public Object[] getAttributes() {
+    return attributes;
+  }
+
+  public PgStructDescriptor getDescriptor() {
+    return descriptor;
+  }
+
+  @Override
+  public Object[] getAttributes(Map<String, Class<?>> map) throws SQLException {
+    Object[] newAttributes = new Object[attributes.length];
+    for (int i = 0; i < attributes.length; i++) {
+      String type = descriptor.pgAttributes()[i].typeName();
+      Class<?> javaType = map.get(type);
+      if (javaType == null || attributes[i] == null) {
+        // I guess if no type is found in the user mapping, we just return the attribute as is?
```

Comment:
Yes - this is documented behavior

> If there is no entry in the given type map that matches the structured type that an attribute represents, the driver uses the standard mapping.


https://docs.oracle.com/javase/8/docs/api/java/sql/Struct.html#getAttributes-java.util.Map-